capstone/capdb/models.py

```diff
@@ -421,13 +458,153 @@ class CaseXML(BaseXMLModel):
+            updated_tree = parsed_updated_case.root
+            original_tree = parsed_orig_case.root
+
+            # If the case is duplicative, just save and move on
+            if updated_tree.find('//duplicative:casebody', nsmap) is not None:
+                super(CaseXML, self).save(force_insert, force_update, *args, **kwargs)
+                return
+
+            if self.case_id == '':
+                self.case_id = parsed_orig_case('case|case').attr('caseid')
+
+            # compare a flat list of xpaths to see if there are structural changes
+            original_xpaths = [original_tree.getelementpath(element) for element in original_tree.iter()]
+            updated_xpaths = [updated_tree.getelementpath(element) for element in updated_tree.iter()]
+            deleted_xpaths = set(original_xpaths) - set(updated_xpaths)
+            added_xpaths = set(updated_xpaths) - set(original_xpaths)
+
+            # iterate over the additions and deletions and add to DM, save element for later
+            # note that elements_to_add_to_tree contains elements from the updated tree while
+            # elements_to_delete_from_tree has elements from the original tree
+            elements_to_delete_from_tree = []
+            for xpath in deleted_xpaths:
+                element = original_tree.find(xpath)
+                # if it's in the casebody, we need to modify the corresponding ALTO
+                if __in_casebody(element):
+                    raise Exception("No current support for removing casebody elements")
+                elements_to_delete_from_tree.append(element)
+
+            elements_to_add_to_tree = []
+            for xpath in added_xpaths:
+                element = updated_tree.find(xpath)
+                if __in_casebody(element):
+                    raise Exception("No current support for adding casebody elements")
+                elements_to_add_to_tree.append(element)
+
+            # now that we've documented additions/deletions, make the tree structures match
+            # here we remove the new elements from the updated tree
+            for element in elements_to_add_to_tree:
+                element.getparent().remove(element)
+
+            # here we remove the deleted elements from the original tree
+            for element in elements_to_delete_from_tree:
+                element.getparent().remove(element)
+
+            # get the alto files associated with the case in the DB
+            alto_files = {}
+            for alto in self.pages.all():
+                alto_fileid = "_".join((["alto"] + alto.barcode.split('_')[1:3]))
+                alto_files[alto_fileid] = alto
+
+            # check to see if any elements in the casebody have been updated so we can update the ALTO
+            # since the tree structures match, we can just iterate over the tree elements and compare values
+            for original_element in original_tree.iter():
```

Comment:
Faster to do this as `original_tree('casebody|casebody')` instead of iterating all and skipping?